src/main/java/org/openpnp/machine/reference/camera/Webcams.java

```diff
@@ -0,0 +1,251 @@
+
+
+
+import org.openpnp.CameraListener;
+import org.openpnp.gui.support.PropertySheetWizardAdapter;
+import org.openpnp.gui.support.Wizard;
+import org.openpnp.gui.wizards.CameraConfigurationWizard;
+import org.openpnp.machine.reference.ReferenceCamera;
+import org.openpnp.machine.reference.camera.wizards.WebcamConfigurationWizard;
+import org.openpnp.spi.PropertySheetHolder;
+import org.openpnp.util.OpenCvUtils;
+import org.simpleframework.xml.Attribute;
+
+import java.awt.Dimension;
+import java.io.File;
+import java.io.IOException;
+
+import javax.imageio.ImageIO;
+
+import com.github.sarxos.webcam.*;
+import com.github.sarxos.webcam.util.jh.JHGrayFilter;
+
+
+
+/**
+ * A Camera implementation based on the OpenCV FrameGrabbers.
+ */
+public class Webcams extends ReferenceCamera implements Runnable , WebcamImageTransformer {
+    
+	@Attribute(required=false)
+	protected String deviceId = "###DEVICE###";
+
+	@Attribute(required=false)
+	private int preferredWidth=0;
+	@Attribute(required=false)
+	private int preferredHeight=0;
+	
+	protected Webcam webcam;
+	private Thread thread;
+	private boolean test;
+	private BufferedImage image;
+
+	private static final JHGrayFilter GRAY = new JHGrayFilter();
+
+
+	@Override
+	public BufferedImage transform(BufferedImage image) {
+		return GRAY.filter(image, null);
+	}
+
+	Webcams() {
+	}
+	
+	@Override
+	public synchronized BufferedImage capture() {
+	    if (thread == null) {
+	        setDeviceId(deviceId);
+	    }
+		if(thread==null) { return null; }
+		try {
+		    BufferedImage img = webcam.getImage();
+		    return transformImage(img);
+		}
+		catch (Exception e) {
+			return null;
+		}
+	}
+	
+	@Override
+    public synchronized void startContinuousCapture(CameraListener listener, int maximumFps) {
+	    if (thread == null) {
+	        setDeviceId(deviceId);
+	    }
+        super.startContinuousCapture(listener, maximumFps);
+    }
+
+    private BufferedImage lastImage=null;
+    private BufferedImage redImage = new BufferedImage(1,1,BufferedImage.TYPE_INT_ARGB);
+					
+
+    public void run() {
+		while (!Thread.interrupted()) {
+			try {
+				BufferedImage image = capture();
+				if (image == null) {
+					image = redImage;
+				}
+					broadcastCapture(image);
+			}
+			catch (Exception e) {
+				e.printStackTrace();
+			}
+			try {
+				Thread.sleep(1000 / 30);
+			}
+			catch (InterruptedException e) {
+				break;
+			}
+		}
+	}
+	
+	public String getDeviceId() {
+		return deviceId;
+	}
+
+	public synchronized void setDeviceId(String deviceId) {
+		this.deviceId = deviceId;
+		if (thread != null) {
+			thread.interrupt();
+			try {
+				thread.join();
+			}
+			catch (Exception e) {
+				e.printStackTrace();
+			}
+			thread = null;
+			webcam.close();
+		}
+		try {
+		webcam=null;
+                for (Webcam cam: Webcam.getWebcams()) {
+                        if(cam.getName().equals(deviceId)) {
+				webcam=cam;
+			}
+		}
+		if(webcam==null) {	// Workaround ???
```

Comment:
Yes, please change it to forceGrayscale. I think it's a useful feature, just a bad variable name.

And thank you for the explanation of the comment. I think the current method is good. If the camera is not found we don't want to confuse the user by opening a different camera. You can remove the comments.